Assets/Scripts/Framework/StateMachine/StateMachine.cs

```diff
@@ -0,0 +1,252 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+using ProefExamen.Framework.Utils;
+
+namespace ProefExamen.Framework.StateMachine
+{
+    /// <summary>
+    /// A class responsible for managing all the states.
+    /// </summary>
+    public class StateMachine : AbstractSingleton<StateMachine>
+    {
+        /// <summary>
+        /// The maximum size of the navigation history.
+        /// </summary>
+        [SerializeField]
+        private int _maxHistorySize = 10;
+
+        /// <summary>
+        /// The registered states.
+        /// </summary>
+        private readonly Dictionary<Type, State> _states = new();
+
+        /// <summary>
+        /// The navigation history.
+        /// </summary>
+        private readonly List<Type> _navigationHistory = new();
+
+        /// <summary>
+        /// Invoked when the state has changed.
+        /// </summary>
+        public Action<State> OnStateChanged;
+
+        /// <summary>
+        /// The current state.
+        /// </summary>
+        public State CurrentState { get; private set; }
+
+        /// <summary>
+        /// The target state.
+        /// </summary>
+        public State TargetState { get; private set; }
+
+        /// <summary>
+        /// The previous state.
+        /// </summary>
+        public State PreviousState
+        {
+            get
+            {
+                if(_navigationHistory.Count < 2)
+                    return null;
+
+                if (!TryGetState(_navigationHistory[^2], out State state))
+                    return null;
```

Comment:
Please place a comment about this weird method of retrieving the second to last index.